sample/sample_proxy.c

```diff
@@ -0,0 +1,529 @@
+/* An optional add-on to the "sample" project.
+ *
+ * This implements a simple picoquic connection-splitting proxy.
+ * The proxy is tied to a single pre-determined server (IP, port) that should
+ * be specified in the CLI.
+ *
+ * Like the sample_server, the proxy listens for incoming connections.
+ *
+ * Upon receipt, the proxy opens a new picoquic connection to its assigned server.
+ * It copies payloads from one socket to the other, performing
+ * necessary decryption and re-encryption in between.
+ *
+ * The connection to the end-to-end connection's "client" (the peer who initiated the
+ * end-to-end connection) is referred to as the "client" connection.
+ * The connection to the backend server that the proxy initiates is the "server" connection.
+ */
+
+#include <stdint.h>
+#include <stdio.h>
+#include <pthread.h>
+#include <assert.h>
+#include <picoquic.h>
+#include <picosocks.h>
+#include <picoquic_utils.h>
+#include <autoqlog.h>
+#include "picoquic_sample.h"
+#include "picoquic_packet_loop.h"
+#include "picoquic_internal.h"
+
+int sample_proxy_callback_to_server(picoquic_cnx_t* cnx,
+    uint64_t stream_id, uint8_t* bytes, size_t length,
+    picoquic_call_back_event_t fin_or_event, void* callback_ctx, void* v_stream_ctx);
+int sample_proxy_callback_to_client(picoquic_cnx_t* cnx,
+    uint64_t stream_id, uint8_t* bytes, size_t length,
+    picoquic_call_back_event_t fin_or_event, void* callback_ctx, void* v_stream_ctx);
+/*
+ * Enable debug print statements.
+ */
+int DEBUG = 1;
```

Comment:
Disable per-packet logging by default? (Though the per-connection logging is probably helpful to leave on.)